text/0003-font-types.md

```diff
@@ -0,0 +1,408 @@
+    let axis_values: Slice<_> = [(-1.0, -1.0), (0., 0.), (0.12, 0.14)]
+        .into_iter()
+        .map(AxisValueMap::from_floats)
+        .collect();
+
+    let segment_maps = Slice::from([axis_values]);
+    let version_major = 1;
+    let version_minor = 0;
+    Avar::new(version_major, version_minor, segment_maps)
+}
+```
+
+This doesn't feel too bad: it can use common Rust idioms and patterns, and is
+not exceptionally verbose. It does have some drawbacks, though: when creating
+objects, they need to be fully described, which makes it verbose, and there
+isn't real mutation; you cannot create a `Slice` and then add or remove items
+afterwards.
+
+It also requires some intermediate allocation: for instance in this example
+storage is allocated for the first `Slice`, those bytes are copied into the
+second `Slice`, and then the whole thing is finally copied into the final table.
+
+There are various ways we could reduce the amount of copying, but they all come
+at the cost of *some* added API complexity.
+
+> **Q**: At what granularity should we think about copying costs? For instance,
+> I suspect that we do not want to copy bytes from tables that are unmodified.
+> Do we want to extend this to not copying bytes from individual subtables that
+> are unmodified, if other parts of a table are? Entries in subtables?
+
+The basic issue here is that if our main types are all just wrappers around raw
+bytes, we expect all of the components of a given object to be encoded
+contiguously; there is no way to say that "this subtable over here is in
+allocation A, and that one is reused from allocation B"; and each case where we
+want to handle that pattern will require some new type. At the very least we can
+imagine a special type for constructing a top-level font object (a header + a
+list of tables), where each table has its own `Buffer` object that may reuse
+storage from elsewhere; and we can then stream those bytes to disk or over the
+network without any additional storage. Doing this for each individual table,
+however, adds additional complexity.
+
+## Plain old data structures
+
+Another very compelling alternative is to just use plain old Rust data
+structures.
+
+In this world, our macros generate *two* types for each 'shape' that we define.
+One of these is the type that we've already seen, (methods for accessing data
+directly from underlying bytes) but the other one is just a plain old Rust
+struct.
+
+We can bikeshed names later; for the purpose of this illustration let's adopt
+the pattern of naming these new types objects `${TYPE}Owned` where `${TYPE}` is
+the name of the previously defined object. So continuing our `avar` example, we
+get:
+
+```rust
+// identical to above
+struct Avar<'a>(..);
+struct SegmentMaps<'a>(..);
+
+struct AvarOwned {
+    pub version_major: u16,
+    pub version_minor: u16,
+    pub segment_maps: Vec<SegmentMapsOwned>,
+}
+
+struct SegmentMapsOwned(pub Vec<AxisValueMap>);
```

Comment:
Is the "Owned" suffix a normal Rust pattern or something unique here? Are there other examples in the Rust ecosystem we might refer to and learn from?

Reply:
It's not unheard of, but I think it's more common when you're retrofitting owned types onto a reference first library which is sort of the case here. The other idiomatic approach would be SegmentMaps for the owned type and SegmentMapsRef (SegmentMapsMut) for the borrowed types(s).

Reply:
I mention above that this name needs bikeshedding. I don't like `Owned`, because it really isn't accurate; this is a different representation. My preference might be to call *this* version `Avar`, and call the read-only version something like `AvarView`, but I don't love that either, and I suspect there's something better out there; for the purpose of this example I didn't want to suddenly change the name of the type we had already introduced.